pkg/controller/syncer/ldap.go

```diff
@@ -0,0 +1,443 @@
+package syncer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"strings"
+
+	legacyconfigv1 "github.com/openshift/api/legacyconfig/v1"
+	userv1 "github.com/openshift/api/user/v1"
+	"github.com/openshift/library-go/pkg/security/ldapclient"
+	"github.com/openshift/library-go/pkg/security/ldaputil"
+	redhatcopv1alpha1 "github.com/redhat-cop/group-sync-operator/pkg/apis/redhatcop/v1alpha1"
+	"github.com/redhat-cop/group-sync-operator/pkg/controller/provider/ldap/builders"
+	ldapbuilders "github.com/redhat-cop/group-sync-operator/pkg/controller/provider/ldap/builders"
+	syncgroups "github.com/redhat-cop/group-sync-operator/pkg/controller/provider/ldap/helpers"
+	"github.com/redhat-cop/group-sync-operator/pkg/controller/provider/ldap/helpers/interfaces"
+	syncerror "github.com/redhat-cop/group-sync-operator/pkg/controller/provider/ldap/helpers/syncerror"
+	"github.com/redhat-cop/operator-utils/pkg/util"
+	"gopkg.in/ldap.v2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	ldapLogger = logf.Log.WithName("syncer_ldap")
+)
+
+type LdapSyncer struct {
+	Name              string
+	GroupSync         *redhatcopv1alpha1.GroupSync
+	Provider          *redhatcopv1alpha1.LdapProvider
+	ReconcilerBase    util.ReconcilerBase
+	CredentialsSecret *corev1.Secret
+	URL               *url.URL
+	CaCertificate     []byte
+	CaCertificateFile string
+	Whitelist         []string
+	Blacklist         []string
+	Syncer            *syncgroups.LDAPGroupSyncer
+}
+
+func (l *LdapSyncer) Init() bool {
+
+	if l.Provider.Whitelist == nil {
+		l.Whitelist = []string{}
+	} else {
+		l.Whitelist = *l.Provider.Whitelist
+	}
+
+	if l.Provider.Blacklist == nil {
+		l.Blacklist = []string{}
+	} else {
+		l.Blacklist = *l.Provider.Blacklist
+	}
+
+	return false
+}
+
+func (l *LdapSyncer) Validate() error {
+	validationErrors := []error{}
+
+	if l.Provider.CredentialsSecret != nil {
+		credentialsSecret := &corev1.Secret{}
+		err := l.ReconcilerBase.GetClient().Get(context.TODO(), types.NamespacedName{Name: l.Provider.CredentialsSecret.Name, Namespace: l.Provider.CredentialsSecret.Namespace}, credentialsSecret)
+
+		if err != nil {
+			validationErrors = append(validationErrors, err)
+		} else {
+
+			// Check that provided secret contains required keys
+			_, usernameSecretFound := credentialsSecret.Data[secretUsernameKey]
+			_, passwordSecretFound := credentialsSecret.Data[secretPasswordKey]
+
+			if !(usernameSecretFound && passwordSecretFound) {
```

Comment:
This only checks if both username and password are missing right? Should we also test if only one or the other is missing?

Reply:
No. This is checking to see whether both the username and password are found. If either are not found, an error would be registered

Reply:
:+1: Looks like I'm just reading this wrong. This should be good to go then.